### PR TITLE
Add cleanup to pools and managers

### PR DESCRIPTION
**Pull request by SethMichaelLarson**
Work for #1100 by adding `.close` for all ConnectionPools and `.clear` for all PoolManagers.

Also noticed that in `TestUtil` there were a few duplicate keys in `parse_url_host_map` so I changed it to a list of tuples to make sure the duplicates weren't silently discarded.

### CODE REVIEW

**Comment by Lukasa** [comment]
Hrm, looks like we have some problems with the queue module here. Want to look into them?

**Comment by SethMichaelLarson** [comment]
@Lukasa Yeah I'll look at them unless you have time today. Won't have much time for open source today.

**Comment by Lukasa** [comment]
There's no rush, it's a weekend. Take time for yourself. =D

**Comment by SethMichaelLarson** [comment]
I think I figured out the `queue` issue with those two tests. Also added `unittest2.TestCase` substitution for all the tests that didn't have the substitution already and required `.addCleanup`. It would be great if you could run your FD test again on this @Lukasa if the CI passes to see if this helps the leaking situation at all? :)

**Comment by SethMichaelLarson** [comment]
So I'm having trouble with GAE as the `import urllib3` comes inside the testcase which is what is causing the `queue` issue but it doesn't seem like I can just move it outside the testcase as then importing `google.appengine.api.prospective_search.prospective_search_stub` from within `nosegae` fails.

File: test/test_util.py [hunk 1]
```diff
@@ -138,34 +139,34 @@ def test_parse_url_normalization(self):
...
+        ('http://google.com/', Url('http', host='google.com', path='/')),
+        ('http://google.com:80/', Url('http', host='google.com', port=80, path='/')),
+        ('http://google.com:80', Url('http', host='google.com', port=80)),
 
         # Auth
-        'http://foo:bar@localhost/': Url('http', auth='foo:bar', host='localhost', path='/'),
-        'http://foo@localhost/': Url('http', auth='foo', host='localhost', path='/'),
-        'http://foo:bar@baz@localhost/': Url('http', auth='foo:bar@baz', host='localhost', path='/'),
-        'http://@': Url('http', host=None, auth='')
-    }
```
Comment (Lukasa) [inline test/test_util.py]: Why are we changing this?

**Comment by Lukasa** [comment]
So *why* is the late import causing the `queue` issue, sorry?

File: test/test_util.py [hunk 2]
```diff
@@ -48,6 +48,7 @@
 # numbers used for timeouts
```
Comment (sigmavirus24) [inline test/test_util.py]: Every change in this file is completely unrelated to the rest of the PR. Please revert these.

File: test/with_dummyserver/test_socketlevel.py [hunk 3]
```diff
@@ -837,9 +856,9 @@ def socket_handler(listener):
                        ':9A:8C:B6:07:CA:58:EE:74:5E')
 
         def request():
+            pool = HTTPSConnectionPool(self.host, self.port,
```
Comment (sigmavirus24) [inline test/with_dummyserver/test_socketlevel.py]: Why is there so much code re-organization in this pull request? It doesn't belong.

File: test/with_dummyserver/test_no_ssl.py [hunk 4]
```diff
@@ -8,21 +8,21 @@
 from dummyserver.testcase import (
         HTTPDummyServerTestCase, HTTPSDummyServerTestCase)
 
+import urllib3
```
Comment (sigmavirus24) [inline test/with_dummyserver/test_no_ssl.py]: This migration should happen in a separate PR

File: test/test_util.py (hunk 1 above)
Comment (SethMichaelLarson) [inline test/test_util.py]: There's two keys with different values for `http://google.com/mail` and `http://google.com/` and that will silently drop one of the test cases.

File: test/with_dummyserver/test_no_ssl.py (hunk 4 above)
Comment (SethMichaelLarson) [inline test/with_dummyserver/test_no_ssl.py]: If this import is inside the testcase there is an error on cleanup.

File: test/test_util.py (hunk 2 above)
Comment (SethMichaelLarson) [inline test/test_util.py]: See my reply to @Lukasa above

**Comment by SethMichaelLarson** [comment]
@Lukasa I don't know why importing urllib3 within the testcase makes the `queue` module within `urllib3` turn to `None` by clean-up time. I'll investigate further.

File: test/test_util.py (hunk 2 above)
Comment (sigmavirus24) [inline test/test_util.py]: There are still spurious and unrelated changes in this file besides turning the dictionary into a list of tuples and removing calls to `dict.items()`

Comment (SethMichaelLarson) [inline test/test_util.py]: As I mentioned above, if you look inside that dictionary there is two duplicated keys with different values within that dictionary so for the entire time the test has been around two of the values were not being tested and silently discarded.

```python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> a = {'a': 1, 'a': 2}
>>> a
{'a': 2}
```

By changing this to a list of tuples it preserves both test cases and changing to a list of tuples necessitates removing the `.items()` the bottom loop where these values are used.

I'm open to moving this to it's own PR but as it stands currently this test is dropping some of it's test cases silently, something I figured we'd want to avoid.

Comment (Lukasa) [inline test/test_util.py]: Ian is saying that there are spurious changes here *beyond* the change to tuple, which he's just OK'd.

Comment (sigmavirus24) [inline test/test_util.py]: That's still unrelated to ensuring a pool manager is closed, but I'm less averse to it being included here for expediency. It also isn't explained in any of the commits or their messages (all the commits are a single line according to the GitHub UI).

So future developers will find this and have no clear understanding of the *why* unless they very closely read the diff. It should at least have been split into its own commit for documenting in our history.

Comment (SethMichaelLarson) [inline test/test_util.py]: The only change beyond the tuple change and the `.items()` change for this file is adding a line of whitespace on L51 and deleting a line of whitespace on L374. Could you point me to what spurious changes in this file you want to see reverted?

File: test/with_dummyserver/test_socketlevel.py (hunk 3 above)
Comment (SethMichaelLarson) [inline test/with_dummyserver/test_socketlevel.py]: This is done because there's a chance that pool is not defined by the time we get outside of the `try-catch` block (eg If an exception occurs during the creation of the `HTTPSConnectionPool`). I can revert this change if that's what's wanted.

File: test/test_util.py (hunk 2 above)
Comment (sigmavirus24) [inline test/test_util.py]: > And I'm sorry for not including the change as a separate commit. I will try to be better.

You can still split it out into its own commit.

> what spurious changes in this file you want to see reverted?

The whitespace changes. They're stylistic and have no bearing on this PR. If you want to submit yet another PR to re-flake8 this project, that's fine, but that should be separate.

Comment (Lukasa) [inline test/test_util.py]: @SethMichaelLarson There's no need to feel bad about not splitting the commit out. Things are overlooked all the time, and that's why we do code review. =) And remember: @sigmavirus24 and I have exactly this kind of experience on the other side of code review all the time (hit me up on Twitter and I'll point out my current OpenSSL code review saga).

Comment (SethMichaelLarson) [inline test/test_util.py]: I'll just revert this whole file and open a separate PR for the dictionary issue. If flaking the unit tests is something that is desired I will put all the work into one PR. I find it difficult to work on the tests and find actual deficiencies when my IDE is lit up like a Christmas tree with linting errors.

Comment (Lukasa) [inline test/test_util.py]: By all means, file a PR with the linting errors. Merging things like that is usually a box-ticking exercise that I can do with half of my brain turned off, so I'm not too worried about it. =D

File: test/with_dummyserver/test_socketlevel.py (hunk 3 above)
Comment (Lukasa) [inline test/with_dummyserver/test_socketlevel.py]: Yeah, this one has to stay.

File: test/test_util.py (hunk 2 above)
Comment (SethMichaelLarson) [inline test/test_util.py]: Okay, as long as it's no trouble. I'll dedicate some time to that task as well as cleaning up this PR.

Comment (Lukasa) [inline test/test_util.py]: No trouble at all.

**Comment by Lukasa** [comment]
So, an FYI: this has become more important in the v2 branch as I refactor to use h11 more closely, because all our connection objects now use a selector and a socket. This doubles the FD use per connection, which means that I can now reproduce the "too many files" error consistently in all tests.

**Comment by SethMichaelLarson** [comment]
I'll get time tonight to work on this branch @Lukasa :) And I'm glad the selectors are useful as well. It would be nice if somehow all connections within a pool could share a selector! :D

**Comment by Lukasa** [comment]
> It would be nice if somehow all connections within a pool could share a selector! :D

So while they *could*, it gets tricky fast: you need to use semaphores to manage the selector because connection pools are thread-safe, and so their shared selector would need to be as well. This is not totally trivial, and incurs a massive performance cost (locking in Python is very slow).

A potential optimisation would be to have one selector per thread in thread-local storage, which would vastly reduce the overhead. That is something we should look into as v2 progresses.

**Comment by SethMichaelLarson** [comment]
Excited to have selectors serve a higher purpose than just waiting for a single socket! :)

**Comment by Lukasa** [comment]
To be clear, the reason we're using selectors for this is twofold. Firstly, and most importantly in terms of user annoyance, they allow us to *finally* consider implementing "total" timeouts in a way that makes sense (that is, "don't wait more than `n` seconds for the entire response").

But the most important reason for me is that they fix a number of problems relating to the fact that the server can respond to us before we're finished uploading. This allows us to send data while we watch for the possibility that the server has already responded. We've not been able to do this before, and it's a huge win. It also opens other use-cases like this, such as support for Expect: 100-continue.

**Comment by SethMichaelLarson** [comment]
@Lukasa I will say the biggest annoyance I have with Python 2.x is that everything that should have a `timeout` parameter (`threading.Lock.acquire`, `subprocess.Popen.communicate`, ...) doesn't, and it makes it really annoying to implement those actual timeouts.

**Comment by Lukasa** [comment]
Hrm, Travis seems to have lost this commit.

**Comment by Lukasa** [comment]
Looks like we have some flake8 problems. =D

**Comment by SethMichaelLarson** [comment]
@Lukasa I'll fix those later today :)

What are your thoughts on leaving the GAE tests as they were? When I pull the import outside the test function then nosegae has trouble importing and when I have the import inside the test function and do cleanup then we run into the `queue` module turning to `None` before cleanup. :/

**Comment by Lukasa** [comment]
We can probably start by leaving the GAE tests as they are for now. Let's make progress as gently as we can. =)

**Comment by SethMichaelLarson** [comment]
Looks like (transient?) failures on Python 3.4 Linux and Python 3.3 OSX. I haven't seen the OSX error before but it looks like a read timeout on localhost?

**Comment by SethMichaelLarson** [comment]
Python 3.3 OSX still failing, another transient issue. :/

**Comment by SethMichaelLarson** [comment]
Yay! Our CI has come through in the end :)

**Comment by sigmavirus24** [comment]
I'm not happy because I think this is still indicative of a problem that can and will occur on a user's system, but I have no objection to the changes in this pull request.

**Comment by Lukasa** [comment]
> I'm not happy because I think this is still indicative of a problem that can and will occur on a user's system, but I have no objection to the changes in this pull request.

So, to be clear, it *absolutely is*. This is a problem we've known about for a very long time, namely that because we pool connections we necessarily keep them alive until we're GC'd and rely on the GC to do that cleaning up.

We could avoid this in some cases by making things context managers, but ultimately that only solves it for well-bounded lifetimes. Otherwise, the problem is present everywhere. However, it is worse here because we create a *bunch* of connections in a tight loop during a test case.
